src/main.rs

```diff
@@ -1,10 +1,64 @@
 use std::env;
+use std::io::{self, Write};
+use std::process;
+use std::borrow::Cow;
 
-fn main() {
-    let arguments = env::args().collect::<Vec<_>>();
-    
-    match arguments.len() {
-        1 => loop { println!("y") },
-        _ => loop { println!("{}", arguments[1]) },
+#[cfg(not(unix))]
+mod platform {
+    use std::ffi::OsString;
+    pub const BUFFER_CAPACITY: usize = 16 * 1024;
+
+    pub fn to_bytes(os_str: OsString) -> Vec<u8> {
+        os_str.into_string().expect("non utf-8 argument only supported on unix").into()
+    }
+}
+
+#[cfg(unix)]
+mod platform {
+    use std::ffi::OsString;
+    pub const BUFFER_CAPACITY: usize = 64 * 1024;
+
+    pub fn to_bytes(os_str: OsString) -> Vec<u8> {
+        use std::os::unix::ffi::OsStringExt;
+        os_str.into_vec()
     }
 }
+
+use platform::*;
+
+fn fill_up_buffer<'a>(buffer: &'a mut [u8], output: &'a [u8]) -> &'a [u8] {
+    if output.len() > buffer.len() / 2 {
+        return output;
+    }
+
+    let mut buffer_size = output.len();
+    buffer[..buffer_size].clone_from_slice(output);
+
+    while buffer_size < buffer.len() {
+        let (left, mut right) = buffer.split_at_mut(buffer_size);
+        right.write(left).unwrap();
```

Comment:
Ideally, I would think you would want to fill up as much of the buffer as possible, to maximize the data written with each syscall.

Either way, you have to be careful to write() only the data that was initialized, or the output would be incorrect.

Wouldn't this all be simpler if you used a pre-allocated `Vec`, or does that add inefficiency?